client/src/common/fileOperations.ts

```diff
@@ -221,6 +225,38 @@ abstract class NotificationFileOperationFeature<I, E extends { readonly files: R
 	protected abstract doSend(event: E, next: (event: E) => void): void;
 }
 
+abstract class CachingNotificationFileOperationFeature<I, E extends { readonly files: ReadonlyArray<I>; }, P> extends NotificationFileOperationFeature<I, E, P> {
+	protected _willListener: code.Disposable | undefined;
+	private _fsPathFileTypes: { [key: string]: code.FileType } = {};
```

Comment:
This is an ever growing cache which we should definitely avoid. When the actual did event fires we should clear the cache. In addition we should make this a normal ES6 Map.

Reply:
Is it guaranteed that these events can't overlap? (eg. clearing the cache in `did` can't impact other concurrent calls to these events?).

Reply:
Since it is all async I guess they theoretical could. But I think it will never happen in practice